crates/ruff_formatter/src/format_element.rs

```diff
@@ -403,6 +399,47 @@ pub trait FormatElements {
     fn end_tag(&self, kind: TagKind) -> Option<&Tag>;
 }
 
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum TextWidth {
+    Width(NonZeroU32),
+    Multiline,
+}
+
+impl TextWidth {
+    pub(crate) const fn new_width(width: u32) -> Self {
+        // SAFETY: 1 + x is guaranteed to be non zero
+        #[allow(unsafe_code)]
+        TextWidth::Width(unsafe { NonZeroU32::new_unchecked(width.saturating_add(1)) })
+    }
+
+    pub fn from_text(text: &str, tab_width: TabWidth) -> TextWidth {
+        let mut width = 0u32;
+
+        for c in text.chars() {
+            let char_width = match c {
+                '\t' => tab_width.value(),
+                '\n' => return TextWidth::Multiline,
+                #[allow(clippy::cast_possible_truncation)]
+                c => c.width().unwrap_or(0) as u32,
```

Comment:
When does `c` not have a width and why would we go with 0 instead, is this about control characters?

Reply:
From the unicode width documentation

> Returns the character's displayed width in columns, or None if the character is a control character other than '\x00'.

So yes, this is about control characters and using 0 seems reasonable for those (this is the same logic as applied by the printer today)